src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -221,93 +221,202 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
-                    // This is equivalent to:
-                    //         unaligned = ((int)pCh % Unsafe.SizeOf<Vector<ushort>>()) / elementsPerByte
-                    //         length = (Vector<ushort>.Count - unaligned) % Vector<ushort>.Count
-                    const int elementsPerByte = sizeof(ushort) / sizeof(byte);
-                    int unaligned = ((int)pCh & (Unsafe.SizeOf<Vector<ushort>>() - 1)) / elementsPerByte;
-                    length = (Vector<ushort>.Count - unaligned) & (Vector<ushort>.Count - 1);
+                    lengthToExamine = UnalignedCountVector128(ref searchSpace);
                 }
-
-            SequentialScan:
-                while (length >= 4)
+            }
+            else if (Vector.IsHardwareAccelerated)
+            {
+                if (length >= Vector<ushort>.Count * 2)
                 {
-                    length -= 4;
+                    lengthToExamine = UnalignedCountVector(ref searchSpace);
+                }
+            }
 
-                    if (pCh[0] == value)
-                        goto Found;
-                    if (pCh[1] == value)
-                        goto Found1;
-                    if (pCh[2] == value)
-                        goto Found2;
-                    if (pCh[3] == value)
-                        goto Found3;
+        SequentialScan:
+            while (lengthToExamine >= 4)
+            {
+                ref char current = ref Add(ref searchSpace, offset);
+                lengthToExamine -= 4;
+
+                if (value == current)
+                    goto Found;
+                if (value == Add(ref current, 1))
+                    goto Found1;
+                if (value == Add(ref current, 2))
+                    goto Found2;
+                if (value == Add(ref current, 3))
+                    goto Found3;
+
+                offset += 4;
+            }
 
-                    pCh += 4;
-                }
+            while (lengthToExamine > 0)
+            {
+                lengthToExamine -= 1;
 
-                while (length > 0)
+                if (value == Add(ref searchSpace, offset))
+                    goto Found;
+
+                offset += 1;
+            }
+
+            // We get past SequentialScan only if IsHardwareAccelerated or intrinsic .IsSupported is true. However, we still have the redundant check to allow
+            // the JIT to see that the code is unreachable and eliminate it when the platform does not have hardware accelerated.
+            if (Avx2.IsSupported)
+            {
+                if (offset < length)
                 {
-                    length--;
+                    if ((((nint)Unsafe.AsPointer(ref searchSpace) + (nint)offset) & (nint)(Vector256<ushort>.Count - 1)) != 0)
+                    {
+                        // Not currently aligned to Vector256 (is aligned to Vector128); this can cause a problem for searches
```

Comment:
> is aligned to Vector128

Not guaranteed, per earlier comments. It's possible that this could result in reads across cache line boundaries (or even page boundaries) in the inner loop. I don't think this would AV because the only caller that this truly matters for is `string.wcslen`, whose input can't be relocated by the GC anyway, but it's good to be aware of.

Reply:
I would have the preference of not making these assumptions or explicitly documenting asserting the case if we do.

People are likely to look at the algorithms/PRs and try to copy the intrinsic code for their own purposes in some cases.